### PR TITLE
Searching OMIM disease by name (partial match)

### PR DESCRIPTION
**Pull request by apraga**
Hi,

From the documentation, I could not find how to look for an HPO entity by name (like `get_hpo_object` in pyhpo ?). Here's a first stab at it. As it's my first commit to a rust project, I could not run the test but would like your advice regarding this feature.

Thanks !

### CODE REVIEW

**Comment by anergictcell** [comment]
Thanks for the suggestion, this makes total sense to add to hpo. I'll try to help a bit with a few comments and remarks:

1. In the PR you mention `searching HPO by name`, but you're actually searching for OmimDiseases. Just a minor comment, but I wasn't sure if that was intentional or not.
2. In the docstring you mention `name / symbol`. I prefer to use `name` for disease or even `substring` for the `disease_by_name` function. `symbol` refers to the name of genes. I am a bit pedantic about the naming :)
3. The `diseases_by_name` function could also return an `Iterator` instead. That way it does not need to allocate new menory, which would be more efficient. You could go the easy way and simply return an`impl Iterator<Item = &OmimDisease>`, but I acutally would prefer a dedicated struct. Something like:

```rust
struct DiseaseIter<'a, F> {
    inner: Filter<Values<'a, OmimDiseaseId, OmimDisease>, F>
}

impl<'a, F: FnMut(&&'a OmimDisease) -> bool + 'a> Iterator for DiseaseIter<'a, F> {
    type Item = &'a OmimDisease;
    fn next(&mut self) -> Option<Self::Item> {
        self.inner.next()
    }
}

impl Ontology {
// ...
    pub fn diseases_by_name<'a>(&'a self, symbol: &'a str) -> DiseaseIter<impl FnMut(&&'a OmimDisease) -> bool + 'a> {
        DiseaseIter{ inner: self.omim_diseases.values().filter(move |disease| disease.name() == symbol)}
    }
// ...
}
```
(I would actually prefer this even more strict and add the `impl FnMut(....` part into the struct itself, but couldn't figure out how to do this right now. However, that's really not important, just a minor nitpick)

Your `disease_by_name` function is great the way it is!

As for testing during development, I use an LSP (rust-analyzer) in my IDE that helps me to make sure I don't make any typing errors while coding.
Then I always run
```bash
# Automatically fix formatting of the source code
cargo fmt

# Show suggestions how to improve my code
cargo clippy

# Run all tests
cargo test
```

Feel free to ask more questions, I'm happy to help whenever I can. And thanks for helping to make `hpo` better!

**Comment by apraga** [comment]
Thanks a lot for your thoughtful and very details answer. I've edited the commit accordingly. However, I was unable to make the following tests works as the result order does not seems to be deterministic.

```
    #[test]
    fn disease_by_name() {
        let ontology = Ontology::from_binary("tests/example.hpo").unwrap();
        let cystinosis = ontology.disease_by_name("Cystinosis").unwrap();
        assert_eq!(cystinosis.name(), "Cystinosis, nephropathic");
    }

    #[test]
    fn diseases_by_name() {
        let ontology = Ontology::from_binary("tests/example.hpo").unwrap();
        let cystinosis: Vec<_> = ontology
            .diseases_by_name("Cystinosis")
            .map(|d| d.name())
            .collect();
        let cystinosis_ref = [
            "Cystinosis, late-onset juvenile or adolescent nephropathic",
            "Cystinosis, adult nonnephropathic",
            "Cystinosis, nephropathic",
        ];
        assert_eq!(cystinosis, cystinosis_ref);
    }
}
```
Actually, I also want to be able to search terms by name too, hence the PR name :)

File: src/ontology.rs [hunk 1]
```diff
@@ -745,6 +758,52 @@ impl Ontology {
         self.omim_diseases.values()
     }
 
+    /// Returns a vector contains diseases whose names contains the provided
```
Comment (anergictcell) [inline src/ontology.rs]: I think you forgot to also update the documentation to clarify that the function is now returning iterator and not a vector

File: src/ontology.rs [hunk 2]
```diff
@@ -745,6 +758,52 @@ impl Ontology {
...
+    /// use hpo::Ontology;
+    /// let ontology = Ontology::from_binary("tests/example.hpo").unwrap();
+    ///
+    /// for result in ontology.diseases_by_name("Cystinosis") {
+    ///     println!("{:?}", result.name());
+    ///  }
+    /// ```
+    pub fn diseases_by_name<'a>(
+        &'a self,
+        symbol: &'a str,
```
Comment (anergictcell) [inline src/ontology.rs]: I suggest to change the variable name from `symbol` to `substring` or `query` to make it more obvious that this is a substring query and not searching for the full name.

File: src/ontology.rs [hunk 3]
```diff
@@ -745,6 +758,52 @@ impl Ontology {
...
+    ) -> DiseaseIter<impl FnMut(&&'a OmimDisease) -> bool + 'a> {
+        DiseaseIter {
+            inner: self
+                .omim_diseases
+                .values()
+                .filter(move |disease| disease.name().contains(symbol)),
+        }
+    }
+
+    /// Returns the first matching diseases whose names contains the provided
```
Comment (anergictcell) [inline src/ontology.rs]: Here you're mixing singular and plural forms, which makes the documentation confusing without looking at the code. I suggest to say `Returns the fist matching disease whose name contains the provided substring`

**Review by anergictcell — COMMENTED** [review]
Hey, the changes look very good, thanks a lot. I entered a few inline comments and have some minor suggestions in addition:


- Can you rebase onto the current `main` branch. Some of the previous changes (e.g. version updates) are appearing in your commit but are already present in the `main` branch.

- Please add some unit test cases as well, not only the doctests. 
- I suggest to add the following tetst cases for both functions:
  - query that does not match any diseases
  - query that matches exactly one disease
  - query that matches several diseases. You rightfully mention that the order changes, so you could only check for the number of returned items or check the the returned disease is within a set of your expected results.

**Comment by apraga** [comment]
The code has been changed according to your suggestions with 3 test cases for both function. Thanks for your help.

**Pull request merged**